plugin/plugin.go

```diff
@@ -0,0 +1,65 @@
+package plugin
+
+import (
+	"fmt"
+
+	"go.uber.org/multierr"
+)
+
+// AddFlags should be used by embedded modules to inject custom flags into `yab`.
+// It adds a set of custom flags with heading `groupName`.
+//
+// The `data` argument should be a pointer to a struct with one field for each
+// command line flag. Each field should use tags like `description` to inform
+// the parser of metadata about the flag. (tags are the same as supported by
+// github.com/jessevdk/go-flags)
+//
+// 	type foo struct {
+// 		Bar string `long:"bar" description:"Sets the 'bar' value."`
+// 	}
+//
+//	AddFlags("Foo Options", "", &foo{})
+//
+// This would inject a custom flag group that looks like:
+//
+//	Foo Options
+// 		--bar	Sets the 'bar' value.
+//
+// In order to retrieve the mutated results of the set flag, users of AddFlags() should
+// retain a reference to the `data` object and check its values after parsing is complete.
+func AddFlags(groupName string, longDescription string, data interface{}) {
+	_flags = append(_flags, &flag{
+		groupName:       groupName,
+		longDescription: longDescription,
+		data:            data,
+	})
+}
+
+// Parser is any object that can add flag groups to itself before performing its parse.
+type Parser interface {
```

Comment:
Any advantage in having this interface not match the `go-flags` interface exactly, and just calling it `FlagParser`? I'm not sure why we have `pluginParserAdapter`, doesn't seem to add much value.

Reply:
It leaks the internals a bit, by exposing go-flags as part of the public API. Figured we'd want to avoid that

I realize the go-flags tags are exposed anyway when users add their options struct, but it's more subtle

Mostly preference, I'll leave the final call to you all

Reply:
Fair enough, I don't have super strong opinions here, so don't mind either way. This seems fine too.

Reply:
Okay I may just leave it as-is in that case, thanks!